### PR TITLE
Improve SCALE encoders for Substrate types

### PR DESCRIPTION
**Pull request by vgeddes**
The SCALE-encoding of substrate calls and various types was rather ad-hoc. Now it is all structured better, improving readability and auditability.

### CODE REVIEW

**Review by alistair-singh — APPROVED** [review]
Looks good! Do we need to add these new libraries to `ethereum/deploy/00-lib.ts` or does hardhat do this for us?

File: ethereum/contracts/SubstrateTypes.sol [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    function MultiAddressWithID(bytes32 account) internal pure returns (bytes memory) {
+        return bytes.concat(bytes1(0x00), account);
+    }
+
+    function H160(address account) internal pure returns (bytes memory) {
+        return abi.encodePacked(account);
+    }
+
+    function None() internal pure returns (bytes memory) {
+        return hex"11";
```
Comment (doubledup) [inline ethereum/contracts/SubstrateTypes.sol:16]: This gets used in place of `bytes1(0x00)` in EthAppPallet and ERC20AppPallet. Should it be 0 instead?

```suggestion
        return hex"00";
```

File: ethereum/contracts/ERC20App.sol [hunk 2]
```diff
@@ -94,23 +96,23 @@ contract ERC20App is AccessControl {
...
+            (bytes memory createCall,) = ERC20AppPallet.create(_token);
             tokens[_token] = true;
             channel.submit(msg.sender, createCall);
         }
 
         bytes memory call;
-        if (_paraId == 0) {
-            call = encodeCall(_token, msg.sender, _recipient, _amount);
+        if (_paraID == 0) {
+            (call,) = ERC20AppPallet.mint(msg.sender, _token, _recipient, _amount);
```
Comment (doubledup) [inline ethereum/contracts/ERC20App.sol:113]: Token and sender aren't swapped in these method calls
```suggestion
            (call,) = ERC20AppPallet.mint(_token, msg.sender, _recipient, _amount);
```

File: ethereum/contracts/ERC20App.sol [hunk 3]
```diff
@@ -94,23 +96,23 @@ contract ERC20App is AccessControl {
...
         } else {
-            call = encodeCallWithParaId(_token, msg.sender, _recipient, _amount, _paraId, _fee);
+            (call,) = ERC20AppPallet.mintAndForward(msg.sender, _token, _recipient, _amount, _paraID, _fee);
```
Comment (doubledup) [inline ethereum/contracts/ERC20App.sol:115]: Token and sender aren't swapped in these method calls
```suggestion
            (call,) = ERC20AppPallet.mintAndForward(_token, msg.sender, _recipient, _amount, _paraID, _fee);
```

**Review by doubledup — COMMENTED** [review]
Looks like the sender & token are swapped in ERC20App, but otherwise looks good.

File: ethereum/contracts/ScaleCodec.sol [hunk 4]
```diff
@@ -105,4 +105,25 @@ library ScaleCodec {
     function encode8(uint8 input) public pure returns (bytes1) {
         return bytes1(input);
     }
+
+    // The biggest value supported by this method is 2^16
+    function encodeUintCompact(uint16 input) internal pure returns (bytes memory) {
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply ethereum/contracts/ScaleCodec.sol:110]: Oops actually this `encodeUintCompact` method wasn't meant to be included in this PR. forgot to remove it.

File: ethereum/contracts/ScaleCodec.sol [hunk 5]
```diff
@@ -105,4 +105,25 @@ library ScaleCodec {
...
+            return abi.encodePacked(uint8(v << 2));
+        } else if (v < 2 ** 14) {
+            return abi.encodePacked(reverse16(uint16(((v << 2) + 1))));
+        } else {
+            return abi.encodePacked(reverse32(uint32(((v << 2) + 2))));
+        }
+    }
+
+    function encodeBytes(bytes memory input) internal pure returns (bytes memory) {
+        require(input.length <= 2**16, "too big");
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply ethereum/contracts/ScaleCodec.sol:123]: Good catch!

File: ethereum/contracts/ScaleCodec.sol [hunk 6]
```diff
@@ -105,4 +105,25 @@ library ScaleCodec {
...
+    // The biggest value supported by this method is 2^16
+    function encodeUintCompact(uint16 input) internal pure returns (bytes memory) {
+        uint256 v = uint256(input);
+
+        if (v < 64) {
+            return abi.encodePacked(uint8(v << 2));
+        } else if (v < 2 ** 14) {
+            return abi.encodePacked(reverse16(uint16(((v << 2) + 1))));
+        } else {
+            return abi.encodePacked(reverse32(uint32(((v << 2) + 2))));
```
[the remark this answers is not in the document]
Comment (vgeddes) [reply ethereum/contracts/ScaleCodec.sol]: Those 2 bits are a tag used for compact variable-length encoding of integers. See here https://docs.substrate.io/reference/scale-codec/#fn-1

**Pull request merged**